File: backend/app/schemas/generation_config.py
"""
生成配置 Schema
Pydantic 模型用于请求/响应验证
"""
from typing import List, Optional
from pydantic import BaseModel, Field, validator
from datetime import datetime


class GenerationConfigBase(BaseModel):
    """生成配置基础模型"""
    
    # 风格层
    tone: str = Field(default="formal", description="语气风格")
    
    # 结构层
    teaching_style: str = Field(default="progressive", description="教学风格")
    content_style: str = Field(default="detailed", description="内容样式")
    
    # 深度层
    difficulty: str = Field(default="intermediate", description="难度等级")
    formula_detail: str = Field(default="derivation", description="公式详细度")
    
    # 受众与格式
    target_audience: str = Field(default="undergraduate", description="目标受众")
    output_format: str = Field(default="lecture", description="输出格式")
    code_language: str = Field(default="python", description="代码语言")
    
    # 内容细节
    chapter_granularity: str = Field(default="standard", description="章节粒度")
    citation_style: str = Field(default="none", description="引用规范")
    interactive_elements: str = Field(default="exercise", description="互动元素")
    
    # 仿真配置
    need_simulation: bool = Field(default=False, description="是否需要仿真")
    simulation_types: List[str] = Field(default_factory=list, description="仿真类型列表")
    
    # 其他
    need_images: bool = Field(default=False, description="是否需要配图")
    
    # 有效值验证
    @validator('tone')
    def validate_tone(cls, v):
        valid = ['formal', 'casual', 'rigorous']
        if v not in valid:
            raise ValueError(f'tone must be one of {valid}')
        return v
    
    @validator('teaching_style')
    def validate_teaching_style(cls, v):
        valid = ['progressive', 'case_driven', 'problem_based', 'comparative']
        if v not in valid:
            raise ValueError(f'teaching_style must be one of {valid}')
        return v
    
    @validator('content_style')
    def validate_content_style(cls, v):
        valid = ['concise', 'detailed', 'visual', 'formula_heavy']
        if v not in valid:
            raise ValueError(f'content_style must be one of {valid}')
        return v
    
    @validator('difficulty')
    def validate_difficulty(cls, v):
        valid = ['beginner', 'intermediate', 'advanced']
        if v not in valid:
            raise ValueError(f'difficulty must be one of {valid}')
        return v
    
    @validator('formula_detail')
    def validate_formula_detail(cls, v):
        valid = ['conclusion_only', 'derivation', 'full_proof']
        if v not in valid:
            raise ValueError(f'formula_detail must be one of {valid}')
        return v
    
    @validator('target_audience')
    def validate_target_audience(cls, v):
        valid = ['undergraduate', 'graduate', 'engineer', 'high_school']
        if v not in valid:
            raise ValueError(f'target_audience must be one of {valid}')
        return v
    
    @validator('output_format')
    def validate_output_format(cls, v):
        valid = ['lecture', 'ppt_outline', 'lab_manual', 'cheatsheet']
        if v not in valid:
            raise ValueError(f'output_format must be one of {valid}')
        return v
    
    @validator('code_language')
    def validate_code_language(cls, v):
        valid = ['python', 'java', 'cpp', 'pseudocode', 'none']
        if v not in valid:
            raise ValueError(f'code_language must be one of {valid}')
        return v
    
    @validator('chapter_granularity')
    def validate_chapter_granularity(cls, v):
        valid = ['brief', 'standard', 'detailed']
        if v not in valid:
            raise ValueError(f'chapter_granularity must be one of {valid}')
        return v
    
    @validator('citation_style')
    def validate_citation_style(cls, v):
        valid = ['none', 'simple', 'academic']
        if v not in valid:
            raise ValueError(f'citation_style must be one of {valid}')
        return v
    
    @validator('interactive_elements')
    def validate_interactive_elements(cls, v):
        valid = ['thinking', 'exercise', 'quiz', 'none']
        if v not in valid:
            raise ValueError(f'interactive_elements must be one of {valid}')
        return v
    
    @validator('simulation_types', each_item=True)
    def validate_simulation_types(cls, v):
        valid = ['animation', 'interactive']
        if v not in valid:
            raise ValueError(f'simulation_types items must be one of {valid}')
        return v


class GenerationConfigCreate(GenerationConfigBase):
    """创建配置请求模型"""
    pass


class GenerationConfigResponse(GenerationConfigBase):
    """配置响应模型"""
    id: int
    user_id: int
    created_at: str  # 返回格式化后的字符串

    class Config:
        from_attributes = True

    @validator('simulation_types', pre=True)
    def parse_simulation_types(cls, v):
        """将数据库字符串转换为列表"""
        if isinstance(v, str):
            if not v:
                return []
            return [t.strip() for t in v.split(',') if t.strip()]
        return v if v else []

    @validator('created_at', pre=True)
    def format_created_at(cls, v):
        """格式化时间为本地时间字符串"""
        if isinstance(v, datetime):
            return v.strftime('%Y-%m-%d %H:%M:%S')
        return v


class GenerationConfigHistory(BaseModel):
    """配置历史列表响应"""
    configs: List[GenerationConfigResponse]
    total: int


class GenerationConfigApply(BaseModel):
    """应用配置到生成任务"""
    config_id: Optional[int] = Field(default=None, description="要应用的配置ID，None表示使用默认")
    course: str = Field(..., description="课程名称")
    knowledge_point: str = Field(..., description="知识点")

File: frontend/src/types/generationConfig.ts
// 生成配置类型定义

// 语气风格
export type Tone = 'formal' | 'casual' | 'rigorous'

// 教学风格
export type TeachingStyle = 'progressive' | 'case_driven' | 'problem_based' | 'comparative'

// 内容样式
export type ContentStyle = 'concise' | 'detailed' | 'visual' | 'formula_heavy'

// 难度等级
export type Difficulty = 'beginner' | 'intermediate' | 'advanced'

// 公式详细度
export type FormulaDetail = 'conclusion_only' | 'derivation' | 'full_proof'

// 目标受众
export type TargetAudience = 'undergraduate' | 'graduate' | 'engineer' | 'high_school'

// 输出格式
export type OutputFormat = 'lecture' | 'ppt_outline' | 'lab_manual' | 'cheatsheet'

// 代码语言
export type CodeLanguage = 'python' | 'java' | 'cpp' | 'pseudocode' | 'none'

// 章节粒度
export type ChapterGranularity = 'brief' | 'standard' | 'detailed'

// 引用规范
export type CitationStyle = 'none' | 'simple' | 'academic'

// 互动元素
export type InteractiveElements = 'thinking' | 'exercise' | 'quiz' | 'none'

// 仿真类型
export type SimulationType = 'animation' | 'interactive'

// 生成配置接口
export interface GenerationConfig {
  id: number
  user_id: number
  
  // 风格层
  tone: Tone
  
  // 结构层
  teaching_style: TeachingStyle
  content_style: ContentStyle
  
  // 深度层
  difficulty: Difficulty
  formula_detail: FormulaDetail
  
  // 受众与格式
  target_audience: TargetAudience
  output_format: OutputFormat
  code_language: CodeLanguage
  
  // 内容细节
  chapter_granularity: ChapterGranularity
  citation_style: CitationStyle
  interactive_elements: InteractiveElements
  
  // 仿真配置
  need_simulation: boolean
  simulation_types: SimulationType[]
  
  // 其他
  need_images: boolean
  
  created_at: string
}

// 创建配置请求
export interface GenerationConfigCreate {
  tone: Tone
  teaching_style: TeachingStyle
  content_style: ContentStyle
  difficulty: Difficulty
  formula_detail: FormulaDetail
  target_audience: TargetAudience
  output_format: OutputFormat
  code_language: CodeLanguage
  chapter_granularity: ChapterGranularity
  citation_style: CitationStyle
  interactive_elements: InteractiveElements
  need_simulation: boolean
  simulation_types: SimulationType[]
  need_images: boolean
}

// 配置历史响应
export interface GenerationConfigHistory {
  configs: GenerationConfig[]
  total: number
}

// 配置选项标签映射
export const configLabels = {
  tone: {
    formal: '正式学术',
    casual: '轻松通俗',
    rigorous: '严谨专业'
  },
  teaching_style: {
    progressive: '循序渐进',
    case_driven: '案例驱动',
    problem_based: '问题导向',
    comparative: '对比分析'
  },
  content_style: {
    concise: '简洁明了',
    detailed: '详细全面',
    visual: '图文并茂',
    formula_heavy: '公式密集'
  },
  difficulty: {
    beginner: '入门',
    intermediate: '进阶',
    advanced: '高级'
  },
  formula_detail: {
    conclusion_only: '仅结论',
    derivation: '推导过程',
    full_proof: '完整证明'
  },
  target_audience: {
    undergraduate: '本科生',
    graduate: '研究生',
    engineer: '工程师',
    high_school: '高中生'
  },
  output_format: {
    lecture: '讲义',
    ppt_outline: 'PPT大纲',
    lab_manual: '实验手册',
    cheatsheet: '速查表'
  },
  code_language: {
    python: 'Python',
    java: 'Java',
    cpp: 'C++',
    pseudocode: '伪代码',
    none: '无代码'
  },
  chapter_granularity: {
    brief: '精简（3-5节）',
    standard: '标准（6-10节）',
    detailed: '详细（10+节）'
  },
  citation_style: {
    none: '无引用',
    simple: '简单标注',
    academic: '学术引用'
  },
  interactive_elements: {
    thinking: '思考题',
    exercise: '练习题',
    quiz: '小测验',
    none: '无互动'
  },
  simulation_types: {
    animation: '动画演示',
    interactive: '交互实验'
  }
}

// 默认配置
export const defaultConfig: GenerationConfigCreate = {
  tone: 'formal',
  teaching_style: 'progressive',
  content_style: 'detailed',
  difficulty: 'intermediate',
  formula_detail: 'derivation',
  target_audience: 'undergraduate',
  output_format: 'lecture',
  code_language: 'python',
  chapter_granularity: 'standard',
  citation_style: 'none',
  interactive_elements: 'exercise',
  need_simulation: false,
  simulation_types: [],
  need_images: false
}
